### PR TITLE
Sbhuller/v2.10.0 elink fix

### LINKED ISSUE #27: FlxCardReader's ELink configuration revamp
Better FELIX card ELink/DMA configuration

- instead of num_links, move to a link_mask list, which elink processors to set up 
- elink_mask config should have (link_id, dma_id) pairs as DMA descriptor is configurable per link 
- update local configuration generator in flxlibs, moving to the changes above, and to generate FlxCardReader modules for reading FW TP links |

### PR DESCRIPTION
**Pull request by ShyamB97**
brranch contains changes to FelixCardReader, I will create a separate branch for the local confgen changes.

### CODE REVIEW

**Comment by bieryAtFnal** [comment]
What is required to review/validate this PR?  Roland is out of the office this week, so it seems like we should request a review from someone else, and if we know what is involved, that can help us understand who we should ask.

**Comment by ShyamB97** [comment]
Hi, the review was requested just to make sure I included all the changes he requested, so I just wanted him to have a quick look at the changes I made to the FelixCardReader. If he is away then we can proceed with the PR.

File: plugins/FelixCardReader.cpp [hunk 1]
```diff
@@ -92,17 +93,25 @@ FelixCardReader::init(const data_t& args)
...
       }
-      auto link_offset = 0;
-      if (linkid >= 5) { // RS FIXME: super ugly... queue names should contain tag for exact elink.
-        link_offset = 5;
-      }
-      auto tag = (linkid - link_offset) * m_elink_multiplier;
-      TLOG_DEBUG(TLVL_WORK_STEPS) << "Creating ElinkModel for target queue: " << target << " elink tag: " << tag;
-      m_elinks[tag] = createElinkModel(qi.inst);
-      if (m_elinks[tag] == nullptr) {
+      //auto link_offset = 0;
```
Comment (jimbrooke) [inline plugins/FelixCardReader.cpp]: is there a reason to leave this code commented, rather than remove it ?

File: plugins/FelixCardReader.cpp [hunk 2]
```diff
@@ -159,13 +172,29 @@ FelixCardReader::do_configure(const data_t& args)
...
     // Configure components
+    TLOG(TLVL_WORK_STEPS) << "Card ID: " << m_card_id;
     TLOG(TLVL_WORK_STEPS) << "Configuring components with Block size:" << m_block_size
                           << " & trailer size: " << m_chunk_trailer_size;
     m_card_wrapper->configure(args);
-    for (unsigned lid = 0; lid < m_num_links; ++lid) {
-      auto tag = lid * m_elink_multiplier;
-      TLOG(TLVL_WORK_STEPS) << "Configuring ElinkHandler with elink tag: " << tag;
-      m_elinks[tag]->set_ids(m_card_id, m_logical_unit, lid, tag);
+    // for (unsigned lid = 0; lid < m_num_links; ++lid) {
```
Comment (jimbrooke) [inline plugins/FelixCardReader.cpp]: same as above - can we delete this commented code ?

**Review by jimbrooke — APPROVED** [review]
Looks OK to me.  I have approved.  Suggest to address the comments above before merging if possible.

File: plugins/FelixCardReader.cpp (hunk 1 above)
Comment (ShyamB97) [reply plugins/FelixCardReader.cpp]: yes we can remove the comments

File: plugins/FelixCardReader.cpp (hunk 2 above)
Comment (ShyamB97) [reply plugins/FelixCardReader.cpp]: see my response above

File: plugins/FelixCardReader.cpp (hunk 1 above)
Comment (ShyamB97) [reply plugins/FelixCardReader.cpp]: done

File: plugins/FelixCardReader.cpp (hunk 2 above)
Comment (ShyamB97) [reply plugins/FelixCardReader.cpp]: done

**Comment by jimbrooke** [comment]
I've added app_confgen.py from sbhuller/confgen-fix to this PR, so we only have to deal with one branch and PR.

**Pull request merged**